addons/portal/controllers/mail.py

```diff
@@ -144,6 +157,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
+            try:
+                document.check_access_rights('read')
+                document.check_access_rule('read')
+            except AccessError:
+                if not access_token or not _check_special_access(model, int(id), token=access_token):
+                    raise Forbidden()
+        except (AccessError, MissingError):
+            raise Forbidden()
+
+        if len(files) > 10:
+            args.append({'error': _("Oops! You can not upload more than 10 files.")})
+            return out % (json.dumps(callback), json.dumps(args))
+
+        Model = request.env['ir.attachment'].sudo()
+
+        for ufile in files:
+            data = ufile.read()
+            file_size = len(data) * 3 / 4  # base64
+            if (file_size / 1024.0 / 1024.0) > 25:
+                args.append({'error': _("Oops! You can not upload a file larger than 25MB.")})
+
+            filename = ufile.filename
+            if request.httprequest.user_agent.browser == 'safari':
+                # Safari sends NFD UTF-8 (where é is composed by 'e' and [accent])
+                # we need to send it the same stuff, otherwise it'll fail
+                filename = unicodedata.normalize('NFD', ufile.filename)
+            try:
+                attachment = Model.create({
+                    'name': filename,
+                    'datas': base64.encodestring(data),
+                    'datas_fname': filename,
+                    'res_model': model,
+                    'res_id': int(id),
+                })
+            except Exception:
+                args.append({'error': _("Something horrible happened")})
+                _logger.exception("Fail to upload attachment %s" % ufile.filename)
+            else:
+                args.append({
+                    'filename': filename,
+                    'mimetype': ufile.content_type,
+                    'id': attachment.id,
+                    'size': attachment.file_size
+                })
+
+        return out % (json.dumps(callback), json.dumps(args))
+
+    @http.route('/portal/content/<int:attachment_id>', type='http', auth="public")
+    def download_attachment(self, attachment_id, model, id, access_token=False, field="datas", filename_field='name', download=None, **kw):
+
+        try:
+            document = request.env[model].browse(int(id))
+            document_sudo = document.sudo().exists()
+            if not document_sudo:
+                raise MissingError("This document does not exist.")
+            try:
+                document.check_access_rights('read')
+                document.check_access_rule('read')
+            except AccessError:
+                if not access_token or not _check_special_access(model, int(id), token=access_token):
+                    raise Forbidden()
+        except (AccessError, MissingError):
+            raise Forbidden()
+
+        record = document_sudo.website_message_ids.mapped('attachment_ids').filtered(lambda x: x.id == int(attachment_id))
```

Comment:
Do you really need to filter them like that? For performance it would be better to make a `search` on attachment for the given `id`, `model` and `attachment_id`. Or make a `browse` with `attachment_id` and then make sure `model` and `id` match.

Reply:
This might be to filter attachments put on notes? Then a check can also be put on `message_type` of `comment`.

Reply:
No it doesn't work, so that's a problem to solve too: the public user can currently see the attachments put on notes by changing the `attachment_id` in the url.

Reply:
field website_message_ids already have domain for 'comment' type messgae.
https://github.com/odoo/odoo/blob/master/addons/portal/models/mail_thread.py#L16